### DISCUSSION TITLE
Cart API is acting weird

### DISCUSSION THREAD

**Discussion by studiodrimmel** [swellstores/community#491] [Help] [answered]
Hi all,

The cart api is broken, and I can't seem to get a good update. Not in my actual store, and not in my developer console in the dashboard.

I have 2 very basic products, they only have a name and a price.
The id's are:
- 65bfc30dedb6e20012fe69a1
- 65bfc3196135f10012ade409

Using the console, I create a new cart session:
<img width="454" alt="Screenshot 2024-02-04 at 19 59 47" src="https://github.com/swellstores/community/assets/101345863/fd377d69-fa10-4610-b981-e27ea15548ba">
This results in the following cart session, with the following id:
```javascript
{
  ...
  "id": "65bfde9a8e4b4a001247ba7c"
}
```

So far so good, right? Next up I do a PUT adding one of the products to the cart using the product_id and a quantity
<img width="651" alt="Screenshot 2024-02-04 at 20 02 18" src="https://github.com/swellstores/community/assets/101345863/92bb2edc-25a6-4721-ae82-8d78fa21aa18">
The result as expected:
```javascript
"items": [
    {
      "product_id": "65bfc30dedb6e20012fe69a1",
      "quantity": 1,
      "id": "65bfdf2c70f651001282486d",
      "price": 1,
      "orig_price": 1,
      "delivery": "shipment",
      "shipment_weight": 0,
      "product_name": "Test product 1",
      "price_total": 1,
      "discount_total": 0,
      "discount_each": 0,
      "tax_total": 0,
      "tax_each": 0
    }
  ],
```

Next up, I want to add another product so therefore I do another put with another product. According to the docs: Updating performs a merge operation. So I perform another PUT with a different product ID:
<img width="595" alt="Screenshot 2024-02-04 at 20 05 25" src="https://github.com/swellstores/community/assets/101345863/f44f6df3-85f3-43f7-9c0b-2f1f566d6ada">
Guess the response?
```javascript
 "items": [
    {
      "product_id": "65bfc3196135f10012ade409",
      "quantity": 1,
      "id": "65bfdf2c70f651001282486d",
      "price": 1,
      "orig_price": 1,
      "delivery": "shipment",
      "shipment_weight": 0,
      "product_name": "Test product 1",
      "price_total": 1,
      "discount_total": 0,
      "discount_each": 0,
      "tax_total": 0,
      "tax_each": 0
    }
  ],
```

As you can see, the actual product_id has changed, but the rest seems to be unchanged. The product name and the price are still the same name and price from the 'original'/first product. I actually expected both products in here since the docs stated it would merge the result. Or atleast a change of the entire object, because I can always merge the arrays myself but this shouldn't be the case in my opinion...

In my project I'm using the node client (since the frontend client uses third party cookies and it's going to be blocked in 2024) and I'm facing the same issues using it (obviously, since the api seems to be broken).

**Comment by ericingram** [comment, marked as answer]
@studiodrimmel You're updating the cart item by changing the product ID and expecting it to reset the product name and price, but the API does not implement that behavior currently.

One can argue that the design should be different, and that is probably a good point - but it's not broken. Array merging the way it works now has caused significant confusion for devs and we might change it in the future with API versioning.

The ideal way to add a new item to the cart is by doing `POST /carts/{id}/items`, which adds an element to the items array. That's what Swell.js does when you call `swell.cart.addItem()`.

Array elements get merged in a PUT request by matching `id` field (if one exists on the model) OR by array index. No other fields like `product_id` are considered normally unless there is some explicit feature logic in the model to do so, but in this case there is not.

**Reply by studiodrimmel** [reply]
Check, thanks for the explanation. Guess the documentation mislead me with the 'merge operation', the /carts/id/items works as expected indeed. I guess looking at swell-js is better than looking at the actual docs on most cases. 
Cheers

**Reply by ericingram** [reply]
The backend API docs are sorely lacking details about specific behaviors for each model, but we have a plan to fill those in soon
